src/MemoryRouter.test.tsx

```diff
@@ -315,6 +315,24 @@ describe("MemoryRouter", () => {
         });
       });
 
+      it("correctly uses default as param", async () => {
+        memoryRouter.setCurrentUrl("/path?queryParam=123");
+        expect(memoryRouter).toMatchObject({
+          asPath: "/path?queryParam=123",
+          pathname: "/path",
+          query: { queryParam: "123" },
+        });
+      });
+
+      it("supports as param", async () => {
```

Comment:
What is Next's behavior when there's path params, like:
`/pages/path/[slug].js`?  If I passed `as = "/path/FOO"`, I'd guess that the result would include `query: { slug: "FOO" }`.  So I'd like to add a test for this too.

Reply:
Hey @scottrippey, thanks for the hint! In fact, we dove a bit deeper into how next router handles cases where both the `href` and the `as` params are given. As it turns out, the `as` params takes precedence of the `href` parameter. For example:

```js
// route = /example/[routeParam]
router.push('/example/foo?queryParam=href', '/example/bar?queryParam=as);
```

Will result in the following router configuration:
```js
{
  pathname: '/example/[routerParam]',
  asPath: '/example/bar?test=as',
  query: {
    routerParam: 'bar',
    queryParam: 'as'
  }
}
```

We will rework the implementation to reflect this behaviour and also add the appropriate tests 🚀

Reply:
Slight correction on the comment from above:

The `as` query params only take precedence if the path of `href` and `as` do not match (i.e. href = "/example/foo", as = "/example/bar"). If the paths match (i.e. href = "/example/foo", as = "/example/foo"), the query params of `href` are used.